src/Api/ResourceClassResolver.php

```diff
@@ -40,33 +40,43 @@ public function __construct(ResourceNameCollectionFactoryInterface $resourceName
      */
     public function getResourceClass($value, string $resourceClass = null, bool $strict = false): string
     {
-        $type = \is_object($value) && !$value instanceof \Traversable ? $this->getObjectClass($value) : $resourceClass;
-        $resourceClass = $resourceClass ?? $type;
+        if ($strict && null === $resourceClass) {
+            throw new InvalidArgumentException('Strict checking is only possible when resource class is specified.');
+        }
+
+        $actualClass = \is_object($value) && !$value instanceof \Traversable ? $this->getObjectClass($value) : null;
+
+        if (null === $actualClass && null === $resourceClass) {
+            throw new InvalidArgumentException('Resource type could not be determined. Resource class must be specified.');
+        }
 
-        if (null === $resourceClass) {
-            throw new InvalidArgumentException(sprintf('No resource class found.'));
+        if (null !== $resourceClass && !$this->isResourceClass($resourceClass)) {
+            throw new InvalidArgumentException(sprintf('Specified class "%s" is not a resource class.', $resourceClass));
         }
 
-        if (
-            null === $type
-            || ((!$strict || $resourceClass === $type) && $isResourceClass = $this->isResourceClass($type))
-        ) {
+        if (null === $actualClass) {
             return $resourceClass;
         }
 
-        // The Resource is an interface
-        if ($value instanceof $resourceClass && $type !== $resourceClass && interface_exists($resourceClass)) {
-            throw new InvalidArgumentException(sprintf('The given object\'s resource is the interface "%s", finding a class is not possible.', $resourceClass));
+        if ($strict && !($typesMatch = is_a($actualClass, $resourceClass, true))) {
+            throw new InvalidArgumentException(sprintf('Object of type "%s" does not match "%s" resource class.', $actualClass, $resourceClass));
+        }
+
+        $mostSpecificResourceClass = null;
+
+        foreach ($this->resourceNameCollectionFactory->create() as $resourceClassName) {
+            if (is_a($actualClass, $resourceClassName, true)) {
+                if (null === $mostSpecificResourceClass || is_subclass_of($resourceClassName, $mostSpecificResourceClass, true)) {
```

Comment:
could be merged with the previous condition to reduce the complexity

Reply:
I've opted to change it to this instead:

```php
foreach ($this->resourceNameCollectionFactory->create() as $resourceClassName) {
    if (!is_a($actualClass, $resourceClassName, true)) {
        continue;
    }

    if (null === $mostSpecificResourceClass || is_subclass_of($resourceClassName, $mostSpecificResourceClass, true)) {
        $mostSpecificResourceClass = $resourceClassName;
    }
}
```